go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,323 @@
+func (q *priorityQueue) Bump(item *qitem) {
+	// Sanity check, necessary because of possible racing between Bump and GC:
+	if item.index < 0 || item.index >= q.n || q.items[item.index].key != item.key {
+		return
+	}
+	item.lat = time.Now()
+	heap.Fix(q, item.index)
+}
+
+const maxConcurrentRequests = 1000
+
+// DiskCache is a local disk LRU CAS and Action Cache cache.
+type DiskCache struct {
+	root             string         // path to the root directory of the disk cache.
+	maxCapacityBytes uint64         // if disk size exceeds this, old items will be evicted as needed.
+	mu               sync.Mutex     // protects the queue.
+	store            sync.Map       // map of keys to qitems.
+	queue            *priorityQueue // keys by last accessed time.
+	sizeBytes        int64          // total size.
+	ctx              context.Context
+	shutdown         chan bool
+	gcTick           uint64
+	gcReq            chan uint64
+	testGcTicks      chan uint64
+}
+
+func New(ctx context.Context, root string, maxCapacityBytes uint64) *DiskCache {
+	res := &DiskCache{
+		root:             root,
+		maxCapacityBytes: maxCapacityBytes,
+		ctx:              ctx,
+		queue: &priorityQueue{
+			items: make([]*qitem, 1000),
+		},
+		gcReq:    make(chan uint64, maxConcurrentRequests),
+		shutdown: make(chan bool),
+	}
+	heap.Init(res.queue)
+	_ = os.MkdirAll(root, os.ModePerm)
+	// We use Git's directory/file naming structure as inspiration:
+	// https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#:~:text=The%20subdirectory%20is%20named%20with%20the%20first%202%20characters%20of%20the%20SHA%2D1%2C%20and%20the%20filename%20is%20the%20remaining%2038%20characters.
+	var wg sync.WaitGroup
+	wg.Add(256)
+	for i := 0; i < 256; i++ {
+		prefixDir := filepath.Join(root, fmt.Sprintf("%02x", i))
+		go func() {
+			defer wg.Done()
+			_ = os.MkdirAll(prefixDir, os.ModePerm)
+			_ = filepath.WalkDir(prefixDir, func(path string, d fs.DirEntry, err error) error {
+				// We log and continue on all errors, because cache read errors are not critical.
+				if err != nil {
+					log.Errorf("Error reading cache directory: %v", err)
+					return nil
+				}
+				if d.IsDir() {
+					return nil
+				}
+				subdir := filepath.Base(filepath.Dir(path))
+				k, err := res.getKeyFromFileName(subdir + d.Name())
+				if err != nil {
+					log.Errorf("Error parsing cached file name %s: %v", path, err)
+					return nil
+				}
+				atime, err := GetLastAccessTime(path)
+				if err != nil {
+					log.Errorf("Error getting last accessed time of %s: %v", path, err)
+					return nil
+				}
+				it := &qitem{
+					key: k,
+					lat: atime,
+				}
+				size, err := res.getItemSize(k)
+				if err != nil {
+					log.Errorf("Error getting file size of %s: %v", path, err)
+					return nil
+				}
+				res.store.Store(k, it)
+				atomic.AddInt64(&res.sizeBytes, size)
+				res.mu.Lock()
+				heap.Push(res.queue, it)
+				res.mu.Unlock()
+				return nil
+			})
+		}()
+	}
+	wg.Wait()
+	go res.gc()
```

Comment:
I think gc should only be called `if len(res.queue) > X && res.sizeBytes > Y && res.queue[0].lat < time.Now().Sub(Z)` to reduce unnecessary processing.